base/gmp.jl

```diff
@@ -915,62 +917,111 @@ function Rational{BigInt}(num::BigInt, den::BigInt)
+end
+
+# note that rationals returned from set_ui and set_si are not checked,
+# set_ui(0, 0) will return 0//0 without errors, just like unsafe_rational
+for (op, T1, T2) in ((:set_ui, Culong, Culong), (:set_si, Clong, Culong))
+    op! = Symbol(op, :!)
+    @eval begin
+        function $op!(z::Rational{BigInt}, a, b)
+            zq = _MPQ(z)
+            ccall($(gmpq(op)), Cvoid, (mpq_t, $T1, $T2), zq, a, b)
+            return sync_rational!(zq)
+        end
+        $op(a, b) = $op!(unsafe_rational(BigInt(), BigInt()), a, b)
+    end
+end
+
+# define add, sub, mul, div, and their inplace versions
+function add!(z::Rational{BigInt}, x::Rational{BigInt}, y::Rational{BigInt})
     if iszero(x.den) || iszero(y.den)
         if iszero(x.den) && iszero(y.den) && isneg(x.num) != isneg(y.num)
             throw(DivideError())
         end
-        return iszero(x.den) ? x : y
+        return set!(z, iszero(x.den) ? x : y)
     end
-    zq = _MPQ()
+    zq = _MPQ(z)
     ccall((:__gmpq_add, :libgmp), Cvoid,
           (mpq_t,mpq_t,mpq_t), zq, _MPQ(x), _MPQ(y))
     return sync_rational!(zq)
 end
-function Base.:-(x::Rational{BigInt}, y::Rational{BigInt})
+
+function sub!(z::Rational{BigInt}, x::Rational{BigInt}, y::Rational{BigInt})
     if iszero(x.den) || iszero(y.den)
         if iszero(x.den) && iszero(y.den) && isneg(x.num) == isneg(y.num)
             throw(DivideError())
         end
-        return iszero(x.den) ? x : -y
+        return set!(z, iszero(x.den) ? x : -y)
     end
-    zq = _MPQ()
+    zq = _MPQ(z)
     ccall((:__gmpq_sub, :libgmp), Cvoid,
           (mpq_t,mpq_t,mpq_t), zq, _MPQ(x), _MPQ(y))
     return sync_rational!(zq)
 end
-function Base.:*(x::Rational{BigInt}, y::Rational{BigInt})
+
+function mul!(z::Rational{BigInt}, x::Rational{BigInt}, y::Rational{BigInt})
     if iszero(x.den) || iszero(y.den)
         if iszero(x.num) || iszero(y.num)
             throw(DivideError())
         end
-        return xor(isneg(x.num),isneg(y.num)) ? -one(BigInt)//zero(BigInt) : one(BigInt)//zero(BigInt)
+        a, b = xor(isneg(x.num),isneg(y.num)) ? (-1, 0) : (1, 0)
+        return set_si!(z, a, b)
     end
-    zq = _MPQ()
+    zq = _MPQ(z)
     ccall((:__gmpq_mul, :libgmp), Cvoid,
           (mpq_t,mpq_t,mpq_t), zq, _MPQ(x), _MPQ(y))
     return sync_rational!(zq)
 end
-function Base.://(x::Rational{BigInt}, y::Rational{BigInt})
+
+function div!(z::Rational{BigInt}, x::Rational{BigInt}, y::Rational{BigInt})
     if iszero(x.den)
         if iszero(y.den)
             throw(DivideError())
         end
-        return isneg(y.num) ? -x : x
+        return set!(z, isneg(y.num) ? -x : x)
     elseif iszero(y.den)
-        return y.den // y.num
+        return set!(z, y.den // y.num)
     elseif iszero(y.num)
         if iszero(x.num)
             throw(DivideError())
         end
-        return (isneg(x.num) ? -one(BigInt) : one(BigInt)) // y.num
+        return set!(z, (isneg(x.num) ? -one(BigInt) : one(BigInt)) // y.num)
```

Comment:
```suggestion
        return set!(z, unsafe_rational(sign(x.num), BigInt()))
```
since we checked just above that `x.num != 0`.

Reply:
What about `return set_si!(z, flipsign(1, x.num), 0))` ?

Reply:
Sure! I don't know the relative performance of `set_si!` vs allocating the `BigInt` first and `set!`, but it's probably equivalent (at least I hope).